### PR TITLE
kie-server: added Form REST services

### PR DESCRIPTION
**Pull request by ibek**
requires this PR to be merged first: https://github.com/droolsjbpm/jbpm/pull/337

### CODE REVIEW

**Comment by ibek** [comment]
this is what you will get by calling the form service: http://pastebin.com/R1MSjCXn

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Variant;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.XML;
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java]: can we use jackson for json related operations? bit concerned of including multiple json libraries...

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    public FormResource(FormServiceBase formServiceBase) {
+        this.formServiceBase = formServiceBase;
+    }
+
+    @GET
+    @Path(PROCESS_FORM_URI)
+    @Produces({ MediaType.APPLICATION_JSON })
+    public Response getProcessForm(@javax.ws.rs.core.Context HttpHeaders headers,
+            @PathParam(CONTAINER_ID) String containerId, @PathParam(PROCESS_ID) String processId,
+            @QueryParam("lang") @DefaultValue("en") String language) {
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java]: what's the support of language parameter?

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/JbpmRestApplicationComponentsService.java [hunk 3]
```diff
@@ -85,12 +102,22 @@
         UserTaskServiceBase userTaskServiceBase = new UserTaskServiceBase(userTaskService, context);
         RuntimeDataServiceBase runtimeDataServiceBase = new RuntimeDataServiceBase(runtimeDataService, context);
         ExecutorServiceBase executorServiceBase = new ExecutorServiceBase(executorService, context);
+        
+        FormServiceBase formServiceBase = new FormServiceBase(definitionService, deploymentService, runtimeDataService, userTaskService, context);
+        Set<FormProvider> formProviders = new HashSet<FormProvider>();
+        
+        InMemoryFormSkeletonProvider iprovider = new InMemoryFormSkeletonProvider();
```
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/JbpmRestApplicationComponentsService.java]: would make sense to use ServiceLoader to find form providers so we are not hardcoding it and allow others to include additional providers if needed

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/FormServiceBase.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        Map<String, String> processData = definitionService.getProcessVariables(containerId, processId);
+
+        if (processData == null) {
+            processData = new HashMap<String, String>();
+        }
+
+        Map<String, Object> renderContext = new HashMap<String, Object>();
+        renderContext.put("process", processDesc);
+        renderContext.put("outputs", processData);
+        renderContext.put("marshallerContext", marshallerHelper);
```
Comment (mswiderski) [inline kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/FormServiceBase.java]: I'd remove the marshaller helper from here and include it in the provider implementation that comes from jbpm code base

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/FormServiceBase.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            if (!StringUtils.isEmpty(template)) {
+                return template;
+            }
+        }
+
+        logger.warn("Unable to find form to render for task '{}' on process '{}'", name, processDesc.getName());
+        return "";
+    }
+
+    protected ContentMarshallerContext getMarshallerContext(String deploymentId, String processId) {
```
Comment (mswiderski) [inline kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/FormServiceBase.java]: move this method to jbpm (as part of the provider impl) as it's internal part of it and while it is unlikely to change it still can :)

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/FormServiceBase.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        DeployedUnit deployedUnit = deploymentService.getDeployedUnit(deploymentId);
+        if (deployedUnit == null) {
+            return new ContentMarshallerContext();
+        }
+        InternalRuntimeManager manager = (InternalRuntimeManager) deployedUnit.getRuntimeManager();
+        return new ContentMarshallerContext(manager.getEnvironment().getEnvironment(), manager.getEnvironment()
+                .getClassLoader());
+    }
+
+    protected ContentMarshallerContext getMarshallerContext(Task task) {
```
Comment (mswiderski) [inline kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/FormServiceBase.java]: same here, move this method to jbpm (as part of the provider impl)

**Comment by mswiderski** [comment]
some tests are needed to make sure we don't have regressions whenever we change things

good piece of work @ibek let's work on it for a while and I am sure it will be included as it does provide lots of value

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java (hunk 1 above)
Comment (ibek) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java]: ok, I will try to rewrite the code which uses the org.json lib.

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java (hunk 2 above)
Comment (ibek) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java]: If look at any label, title, error message in the form file.... it looks like this <property name="label" value="quot;SSClassquot;,quot;quot;quot;SSStylequot;,quot;quot;quot;caquot;,quot;quot;quot;dequot;,quot;quot;quot;enquot;,quot;Agequot;quot;esquot;,quot;quot;quot;frquot;,quot;quot;quot;jaquot;,quot;quot;quot;ptquot;,quot;quot;quot;zhquot;,quot;quot;"/>

The purpose of this parameter is to filter these properties, so they would contain only a text representing the value. In this example after filtering the property would look like:

<property name="label" value="Age"/>

which is much easier to work with.

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java (hunk 1 above)
Comment (ibek) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java]: can jackson transform xml string to json object?

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/JbpmRestApplicationComponentsService.java (hunk 3 above)
Comment (ibek) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/JbpmRestApplicationComponentsService.java]: I am not sure how this should work. The problem is that the REST form service depends on the InMemoryFormSkeletonProvider.

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/FormServiceBase.java (hunk 4 above)
Comment (ibek) [inline kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/FormServiceBase.java]: ok, this is copy paste from some another service base

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/JbpmRestApplicationComponentsService.java (hunk 3 above)
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/JbpmRestApplicationComponentsService.java]: isn't it enough to depend on FormProvider interface only? Couldn't see any additional methods in the skeleton provider besides those coming from interface.

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java (hunk 1 above)
Comment (mswiderski) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java]: frankly speaking no idea though worth looking into, especially if we want to support not only json but jaxb and xstream as well

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/JbpmRestApplicationComponentsService.java (hunk 3 above)
Comment (ibek) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/JbpmRestApplicationComponentsService.java]: I mean that the REST form service depends on the output of InMemoryFormSkeletonProvider ... so the structure is known to the service. I cannot imagine there would be any other provider than InMemoryFormSkeletonProvider.

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java (hunk 1 above)
Comment (ibek) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/FormResource.java]: do we really need jaxb and xstream? ... only json seems to me as a good format for forms .... the problem is that we don't have an object of the form, so we could transform the object to XML or JSON on demand.

**Comment by mswiderski** [comment]
@ibek closing this as we discussed, there will be new PR that builds this support around KIE Server extensions that incorporates this work in great part, thanks a lot

**Pull request closed without merging**